cpp/src/arrow/util/map.h

```diff
@@ -17,13 +17,21 @@
 
 #pragma once
 
+#include <type_traits>
 #include <utility>
 
 #include "arrow/result.h"
 
 namespace arrow {
 namespace internal {
 
+/// Functor to make enums hashable.
+template <typename T, typename R = typename std::underlying_type<T>::type,
+          typename = std::enable_if<std::is_enum<T>::value>>
+struct EnumHash {
+  R operator()(T val) const { return static_cast<R>(val); }
+};
```

Comment:
Is it actually used in this PR?

Reply:
It was used for `std::unordered_map<RoundMode, ...>` but this map was removed and converted into a switch statement inside of a function. Also, it is used in the `scalar_arithmetic_test.cc` but it will not be used as maps will be converted to `std::vector<std::pair<...>>`.